pkg/remotecluster/remotecluster.go

```diff
@@ -0,0 +1,202 @@
+)
+
+type Event struct {
+	Cluster *RemoteCluster
+	Type    EventType
+	ObjType ObjectType
+	Objs    []interface{}
+}
+
+func New(clusterID string, clientSet kubernetes.Interface) *RemoteCluster {
+
+	factory := informers.NewSharedInformerFactory(clientSet, time.Hour*24)
+	podInformer := factory.Core().V1().Pods().Informer()
+	networkPolicyInformer := factory.Networking().V1().NetworkPolicies().Informer()
+
+	resourceWatcher := &RemoteCluster{
+		stopCh:                make(chan struct{}),
+		ClusterID:             clusterID,
+		ClientSet:             clientSet,
+		podInformer:           podInformer,
+		networkPolicyInformer: networkPolicyInformer,
+		eventChanMutex:        &sync.Mutex{},
+	}
+
+	podInformer.AddEventHandler(resourceWatcher)
+	networkPolicyInformer.AddEventHandler(resourceWatcher)
+
+	return resourceWatcher
+}
+
+func (rw *RemoteCluster) HasSynced() bool {
+	return rw.podInformer.HasSynced() &&
+		rw.networkPolicyInformer.HasSynced()
+}
+
+// Stop will stop the running informers
+func (rw *RemoteCluster) Stop() {
+	close(rw.stopCh)
+}
+
+func (rw *RemoteCluster) Stopped() bool {
+	select {
+	case _, ok := <-rw.stopCh:
+		// channel is not ok to read, means it's closed now
+		return !ok
+	default:
+		// channel was ok to read, but had no messages
+		return false
+	}
+}
+
+func (rw *RemoteCluster) Run(onSyncDoneFunc func(resourceWatcher *RemoteCluster)) {
+	go rw.podInformer.Run(rw.stopCh)
+	go rw.networkPolicyInformer.Run(rw.stopCh)
+
+	go func() {
+		if !cache.WaitForCacheSync(rw.stopCh, rw.podInformer.HasSynced) {
+			klog.Warning("Timed out waiting for pod informer to sync")
+		}
+
+		if !cache.WaitForCacheSync(rw.stopCh, rw.networkPolicyInformer.HasSynced) {
+			klog.Warning("Timed out waiting for NetworkPolicy informer to sync")
+		}
+
+		if onSyncDoneFunc != nil {
+			onSyncDoneFunc(rw)
+		}
+	}()
+}
+
+func (rw *RemoteCluster) GetPods() []interface{} {
+	return rw.podInformer.GetStore().List()
+}
+
+func (rw *RemoteCluster) GetNetworkPolicies() []interface{} {
+	return rw.networkPolicyInformer.GetStore().List()
+}
+
+func (rw *RemoteCluster) SetEventChannel(eventChan chan *Event) {
+	rw.eventChanMutex.Lock()
+	rw.eventChan = eventChan
+	rw.eventChanMutex.Unlock()
+}
+
+func (rw *RemoteCluster) enqueueEvent(event *Event) {
+
+	// lock used as memory barrier to make sure eventChannel ref
+	// is synchronized between threads
+	rw.eventChanMutex.Lock()
+	if rw.eventChan != nil {
+		rw.eventChan <- event
+	}
+	rw.eventChanMutex.Unlock()
+}
+
+func (rw *RemoteCluster) OnAdd(obj interface{}) {
+
+	var objType ObjectType
+
+	if pod, ok := obj.(*v1.Pod); ok {
+		klog.Infof("Pod discovered in %s: %s", rw.ClusterID, pod.GetSelfLink())
+		objType = Pod
+	} else if np, ok := obj.(*v1net.NetworkPolicy); ok {
+		klog.Infof("NetworkPolicy discovered in %s: %s", rw.ClusterID, np.GetSelfLink())
+		objType = NetworkPolicy
+	} else {
+		klog.Errorf("OnAdd for unexpected object type: %v", obj)
+		return
+	}
+
+	rw.enqueueEvent(&Event{
```

Comment:
I don't want to make that more complex now,  the workqueue does not help us here.

I'm trying to give all the producers the channel to drop things in, I could give it to them beforehand and have may be some sort of signal to let them start pushing events (instead of channel/no channel as it is now).

It's likely I will get rid of the SetEventChannel thing during the real implementation phase, but I think it's ok for now with the lock.

Reply:
Actually it would simplify things by a lot. But as you wish :)